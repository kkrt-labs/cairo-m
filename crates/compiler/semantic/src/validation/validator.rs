//! # Validator Trait and Registry
//!
//! This module defines the trait for semantic validators and provides
//! a registry system for organizing multiple validators.
//!
//! # Architecture
//!
//! The validator system uses a plugin-like architecture where different
//! validation concerns are separated into individual validators. This allows
//! for:
//! - Modular validation logic
//! - Easy addition of new validation rules
//! - Configurable validation passes
//! - Independent testing of validation rules
//!
//! # Usage
//!
//! ```rust,ignore
//! // Create a registry with default validators
//! let registry = create_default_registry();
//!
//! // Or build a custom registry
//! let registry = ValidatorRegistry::new()
//!     .add_validator(ScopeValidator)
//!     .add_validator(TypeValidator);  // TODO: Implement
//!
//! // Run validation
//! let diagnostics = registry.validate_all(&db, file, &index);
//! ```

use crate::db::SemanticDb;
use crate::validation::diagnostics::{Diagnostic, DiagnosticCollection};
use crate::{File, SemanticIndex};

/// Trait for semantic validators
///
/// Each validator implements a specific category of semantic analysis,
/// such as scope checking, type checking, or control flow analysis.
///
/// # Implementation Guidelines
///
/// - Validators should be stateless and thread-safe
/// - Each validator should focus on a single concern
/// - Validators should not modify the semantic index
/// - Return diagnostics in order of source location when possible
pub trait Validator {
    /// Validate the semantic index and return diagnostics
    ///
    /// This is the main entry point for validation logic. Implementers
    /// should analyze the provided semantic index and return any issues
    /// found as a vector of diagnostics.
    ///
    /// # Parameters
    ///
    /// - `db`: Database for additional queries if needed
    /// - `file`: The file being validated (for context)
    /// - `index`: The semantic index containing all semantic information
    ///
    /// # Returns
    ///
    /// A vector of diagnostics representing any issues found during validation.
    fn validate(&self, db: &dyn SemanticDb, file: File, index: &SemanticIndex) -> Vec<Diagnostic>;

    /// Get the name of this validator (for debugging/logging)
    ///
    /// By default, this returns the type name, but implementers can
    /// override it to provide more descriptive names.
    fn name(&self) -> &'static str {
        std::any::type_name::<Self>()
    }

    // TODO: Add more trait methods for advanced validator features:
}

/// Registry for managing multiple validators
///
/// The registry maintains a collection of validators and provides
/// methods to run them collectively. It handles:
/// - Aggregating diagnostics from multiple validators
/// - Managing validator lifecycle
/// - Providing a unified interface for validation
///
/// TODO: Add support for:
/// - Validator ordering/dependencies
/// - Parallel validation execution
/// - Validator configuration and filtering
/// - Performance metrics collection
#[derive(Default)]
pub struct ValidatorRegistry {
    validators: Vec<Box<dyn Validator>>,
}

impl ValidatorRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Add a validator to the registry
    pub fn add_validator<V: Validator + 'static>(mut self, validator: V) -> Self {
        self.validators.push(Box::new(validator));
        self
    }

    /// Run all validators and collect diagnostics
    pub fn validate_all(
        &self,
        db: &dyn SemanticDb,
        file: File,
        index: &SemanticIndex,
    ) -> DiagnosticCollection {
        let mut collection = DiagnosticCollection::new();

        for validator in &self.validators {
            let diagnostics = validator.validate(db, file, index);
            collection.extend(diagnostics);
        }

        collection
    }

    /// Get the number of registered validators
    pub fn len(&self) -> usize {
        self.validators.len()
    }

    /// Check if the registry is empty
    pub fn is_empty(&self) -> bool {
        self.validators.is_empty()
    }
}

/// Create a default validator registry with basic semantic validators
///
/// This provides a sensible default set of validators for most use cases.
/// Currently includes:
/// - **ScopeValidator**: Undeclared variables, unused variables, duplicate definitions
/// - **FunctionCallValidator**: Validate function call arity and argument types
///
/// TODO: Expand default registry with additional validators:
/// - **StructFieldValidator**: Validate field access on struct types (p.z should error)
/// - **StructLiteralValidator**: Validate struct literal completeness and field names
/// - **IndexingValidator**: Validate indexing on non-array types (num[0] should error)
/// - **AssignmentValidator**: Validate assignment compatibility and mutability
/// - **ReturnValidator**: Validate return type consistency and placement
/// - **TypeValidator**: General type checking and compatibility
/// - **ControlFlowValidator**: Reachability analysis, dead code detection
/// - **RecursiveTypeValidator**: Detect recursive struct definitions without indirection
/// - **ConstExpressionValidator**: Validate const expressions are compile-time evaluable
/// - **ModuleValidator**: Import/export validation, module resolution
/// - **StyleValidator**: Code style and best practices
/// - **SecurityValidator**: Security-related checks
/// - **PerformanceValidator**: Performance hints and optimizations
pub fn create_default_registry() -> ValidatorRegistry {
    ValidatorRegistry::new()
        .add_validator(crate::validation::scope_check::ScopeValidator)
        .add_validator(crate::validation::function_call_validator::FunctionCallValidator)
        .add_validator(crate::validation::struct_field_validator::StructFieldValidator)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Mock validator for testing
    struct MockValidator {
        name: &'static str,
        diagnostics: Vec<Diagnostic>,
    }

    impl MockValidator {
        fn new(name: &'static str, diagnostics: Vec<Diagnostic>) -> Self {
            Self { name, diagnostics }
        }
    }

    impl Validator for MockValidator {
        fn validate(
            &self,
            _db: &dyn SemanticDb,
            _file: File,
            _index: &SemanticIndex,
        ) -> Vec<Diagnostic> {
            self.diagnostics.clone()
        }

        fn name(&self) -> &'static str {
            self.name
        }
    }

    #[test]
    fn test_validator_registry() {
        let span = chumsky::span::SimpleSpan::from(0..4);
        let validator1 =
            MockValidator::new("test1", vec![Diagnostic::undeclared_variable("var1", span)]);
        let validator2 =
            MockValidator::new("test2", vec![Diagnostic::unused_variable("var2", span)]);

        let registry = ValidatorRegistry::new()
            .add_validator(validator1)
            .add_validator(validator2);

        assert_eq!(registry.len(), 2);
        assert!(!registry.is_empty());
    }
}
