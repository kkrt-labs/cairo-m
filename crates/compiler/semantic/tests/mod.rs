//! # Semantic Validation Tests
//!
//! This module contains comprehensive tests for semantic validation organized by concern.
//! Tests are structured to provide clear visibility into what semantic features are
//! implemented and validated.
//!
//! ## Test Organization
//!
//! - `scoping/` - Variable scoping, visibility, and declaration tests
//! - `types/` - Type checking, type resolution, and type validation tests
//! - `control_flow/` - Control flow analysis, unreachable code, missing returns
//! - `functions/` - Function calls, parameter validation, return type checking
//! - `structures/` - Struct definition, field access, and struct validation
//! - `expressions/` - Expression validation, operator usage, literal validation
//! - `statements/` - Statement validation (let, assignment, etc.)
//! - `integration/` - End-to-end integration tests with complex scenarios
//!
//! ## Test Utilities
//!
//! This module re-exports the core testing utilities:
//! - `assert_semantic_ok!(code)` - Assert code validates without errors
//! - `assert_semantic_err!(code)` - Assert code produces validation errors
//! - `assert_diagnostics_snapshot!(file, name)` - Snapshot test for .cm files

use cairo_m_compiler_diagnostics::{
    build_diagnostic_message, DiagnosticCode, DiagnosticCollection,
};
use cairo_m_compiler_parser::{Db as ParserDb, SourceProgram, Upcast};
use cairo_m_compiler_semantic::validation::validator::create_default_registry;
use cairo_m_compiler_semantic::{semantic_index::semantic_index, SemanticDb};
use insta::assert_snapshot;

#[salsa::db]
#[derive(Clone, Default)]
pub struct TestDb {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for TestDb {}
#[salsa::db]
impl ParserDb for TestDb {}
#[salsa::db]
impl SemanticDb for TestDb {}

impl Upcast<dyn ParserDb> for TestDb {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

fn test_db() -> TestDb {
    TestDb::default()
}

/// Run semantic validation on source code
fn run_validation(source: &str, file_name: &str) -> DiagnosticCollection {
    let db = test_db();
    let source_program = SourceProgram::new(&db, source.to_string(), file_name.to_string());

    // Build semantic index
    let index = semantic_index(&db, source_program)
        .as_ref()
        .expect("Got unexpected parse errors");

    // Run validation
    let registry = create_default_registry();
    registry.validate_all(&db, source_program, index)
}

// Format diagnostics for snapshot testing using ariadne for beautiful error reports
fn format_diagnostics_for_snapshot(
    diagnostics: &DiagnosticCollection,
    source: &str,
    fixture_name: &str,
) -> String {
    let mut result = String::new();

    // Add header
    result.push_str(&format!("Fixture: {fixture_name}\n"));
    result.push_str(&"=".repeat(60));
    result.push('\n');
    result.push_str(&format!("Source code:\n{source}\n"));
    result.push_str(&"=".repeat(60));
    result.push('\n');

    if diagnostics.is_empty() {
        result.push_str("No diagnostics found.\n");
        return result;
    }

    result.push_str(&format!("Found {} diagnostic(s):\n\n", diagnostics.len()));

    // Create ariadne reports for each diagnostic
    for (i, diagnostic) in diagnostics.all().iter().enumerate() {
        let message = build_diagnostic_message(source, diagnostic, false);
        result.push_str(&format!("--- Diagnostic {} ---\n", i + 1));
        result.push_str(&message);
        if i < diagnostics.len() - 1 {
            result.push('\n');
        }
    }

    result
}

/// Assert that inline code validates successfully without any diagnostics
/// Similar to parser's assert_parses_ok! but for semantic validation
#[track_caller]
pub fn assert_semantic_ok_impl(code: &str, test_name: &str) {
    assert_semantic_ok_impl_with_options(code, test_name, false)
}

/// Assert that inline code validates successfully without any diagnostics
/// with option to mute unused variable warnings
#[track_caller]
pub fn assert_semantic_ok_impl_with_options(
    code: &str,
    test_name: &str,
    show_unused_warnings: bool,
) {
    let diagnostics = run_validation(code, test_name);
    let filtered_diagnostics = if show_unused_warnings {
        diagnostics
    } else {
        filter_unused_variable_warnings(&diagnostics)
    };

    if !filtered_diagnostics.is_empty() {
        let report = format_diagnostics_for_snapshot(&filtered_diagnostics, code, test_name);
        panic!("Expected successful semantic validation, but got diagnostics:\n{report}");
    }
}

/// Assert that inline code fails semantic validation and produces diagnostics
/// Similar to parser's assert_parses_err! but for semantic validation
#[track_caller]
pub fn assert_semantic_err_impl(code: &str, test_name: &str) {
    assert_semantic_err_impl_with_options(code, test_name, false)
}

/// Assert that inline code fails semantic validation and produces diagnostics
/// with option to mute unused variable warnings
#[track_caller]
pub fn assert_semantic_err_impl_with_options(
    code: &str,
    test_name: &str,
    show_unused_warnings: bool,
) {
    let diagnostics = run_validation(code, test_name);
    let filtered_diagnostics = if show_unused_warnings {
        diagnostics
    } else {
        filter_unused_variable_warnings(&diagnostics)
    };

    if filtered_diagnostics.is_empty() {
        panic!("Expected semantic validation to fail, but it succeeded without diagnostics.");
    }

    let snapshot_content = format_diagnostics_for_snapshot(&filtered_diagnostics, code, test_name);

    // Extract local test name for better snapshot organization
    let base_path = "semantic_tests::";
    let local_test_name = test_name.split(base_path).nth(1).unwrap_or(test_name);

    insta::with_settings!({
        description => format!("Inline semantic validation error test: {}", local_test_name),
        omit_expression => true,
        sort_maps => true,
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(format!("diagnostics__{}", local_test_name), snapshot_content);
    });
}

/// Helper macro for snapshot testing
#[macro_export]
macro_rules! assert_diagnostics_snapshot {
    ($fixture:expr, $snapshot_name:expr) => {
        $crate::assert_diagnostics_snapshot($fixture, $snapshot_name);
    };
    ($fixture:expr, $snapshot_name:expr, $description:expr) => {
        $crate::validation::tests::assert_diagnostics_snapshot_with_context(
            $fixture,
            $snapshot_name,
            Some($description),
            None::<&dyn std::fmt::Debug>,
        );
    };
    ($fixture:expr, $snapshot_name:expr, $description:expr, $context:expr) => {
        $crate::validation::tests::assert_diagnostics_snapshot_with_context(
            $fixture,
            $snapshot_name,
            Some($description),
            Some($context),
        );
    };
}

/// Helper macro for clean fixture testing
#[macro_export]
macro_rules! test_fixture_clean {
    ($fixture:expr) => {
        $crate::test_fixture_clean($fixture);
    };
}

/// Macro to assert that inline code validates successfully
#[macro_export]
macro_rules! assert_semantic_ok {
    ($code:expr) => {{
        let function_name = stdext::function_name!();
        $crate::assert_semantic_ok_impl($code, function_name)
    }};
    ($code:expr, show_unused) => {{
        let function_name = stdext::function_name!();
        $crate::assert_semantic_ok_impl_with_options($code, function_name, true)
    }};
}

/// Macro to assert that inline code fails semantic validation
#[macro_export]
macro_rules! assert_semantic_err {
    ($code:expr) => {{
        let function_name = stdext::function_name!();
        $crate::assert_semantic_err_impl($code, function_name)
    }};
    ($code:expr, show_unused) => {{
        let function_name = stdext::function_name!();
        $crate::assert_semantic_err_impl_with_options($code, function_name, true)
    }};
}

// ===== Helper functions for test code generation =====

/// Helper to wrap statement code inside a function, since most statements are not top-level.
pub fn in_function(code: &str) -> String {
    format!("func test() {{ {code} }}")
}

/// Helper to wrap code in a function with a return type
pub fn in_function_with_return(code: &str, return_type: &str) -> String {
    format!("func test() -> {return_type} {{ {code} }}")
}

/// Helper to wrap code in a function with parameters
pub fn in_function_with_params(code: &str, params: &str) -> String {
    format!("func test({params}) {{ {code} }}")
}

/// Helper to wrap code in a function with both parameters and return type
pub fn in_function_with_params_and_return(code: &str, params: &str, return_type: &str) -> String {
    format!("func test({params}) -> {return_type} {{ {code} }}")
}

/// Helper to create a struct definition with the given fields
pub fn with_struct(struct_name: &str, fields: &str, code: &str) -> String {
    format!("struct {struct_name} {{ {fields} }}\n\n{code}")
}

/// Helper to create multiple function definitions
pub fn with_functions(functions: &str, main_code: &str) -> String {
    format!("{functions}\n\n{main_code}")
}

// ===== Diagnostic filtering functions =====

/// Filter out unused variable warnings from diagnostics
fn filter_unused_variable_warnings(diagnostics: &DiagnosticCollection) -> DiagnosticCollection {
    let filtered: Vec<_> = diagnostics
        .all()
        .iter()
        .filter(|d| d.code != DiagnosticCode::UnusedVariable)
        .cloned()
        .collect();
    DiagnosticCollection::from(filtered)
}

// Test modules organized by concern
pub mod control_flow;
pub mod expressions;
pub mod functions;
pub mod integration;
pub mod scoping;
pub mod statements;
pub mod structures;
pub mod types;
