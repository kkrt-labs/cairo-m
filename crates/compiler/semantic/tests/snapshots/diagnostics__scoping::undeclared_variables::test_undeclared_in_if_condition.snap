---
source: crates/compiler/semantic/tests/mod.rs
description: "Inline semantic validation error test: scoping::undeclared_variables::test_undeclared_in_if_condition"
---
Fixture: semantic_tests::scoping::undeclared_variables::test_undeclared_in_if_condition
============================================================
Source code:

        fn test() {
            if (undefined_condition) {
                let x = 1;
            }
        }
    
============================================================
Found 2 diagnostic(s):

--- Diagnostic 1 ---
[1001] Error: Undeclared variable 'undefined_condition'
   ╭─[ semantic_tests::scoping::undeclared_variables::test_undeclared_in_if_condition:3:17 ]
   │
 3 │             if (undefined_condition) {
   │                 ─────────┬─────────  
   │                          ╰─────────── Undeclared variable 'undefined_condition'
───╯

--- Diagnostic 2 ---
[3002] Error: Function 'test' doesn't return on all paths
   ╭─[ semantic_tests::scoping::undeclared_variables::test_undeclared_in_if_condition:2:12 ]
   │
 2 │         fn test() {
   │            ──┬─  
   │              ╰─── Function 'test' doesn't return on all paths
───╯
